content/blog/2022/2022-05-27-one-stop-guide-mapstruct.md

```diff
@@ -0,0 +1,2116 @@
+    BasicUserDTO convert(BasicUser user);
+
+    @InheritInverseConfiguration
+    BasicUser convert(BasicUserDTO userDTO);
+}
+```
+
+This can be used for straightforward mappings between entity and DTO.
+
+### Exception Handling during Mapping
+
+Exceptions are unavoidable, hence, MapStruct provides support to handle exceptions by making the life of developers quite easy. Let’s say if we want to validate the id field for any invalid values, then we can define a utility class named as `Validator` :
+
+```java
+public class Validator {
+    public int validateId(int id) throws ValidationException {
+        if(id == -1){
+            throw new ValidationException("Invalid ID value");
+        }
+        return id;
+    }
+}
+```
+
+Next, we can define an exception class, `ValidationException` which we will use in our mapper:
+
+```java
+public class ValidationException extends RuntimeException {
+
+    public ValidationException(String message, Throwable cause) {
+        super(message, cause);
+    }
+
+    public ValidationException(String message) {
+        super(message);
+    }
+}
+```
+
+Finally, we will update our `UserMapper` by including the `Validator` class and throw `ValidationException` wherever we are mapping the id fields:
+
+```java
+@Mapper(uses = {ManagerMapper.class, Validator.class})
+public interface UserMapper {
+    UserMapper INSTANCE = Mappers.getMapper(UserMapper.class);
+    BasicUserDTO convert(BasicUser user) throws ValidationException;
+
+    @InheritInverseConfiguration
+    BasicUser convert(BasicUserDTO userDTO) throws ValidationException;
+    ...
+}
+```
+
+The implementation class after generation would look something like below:
+
+```java
+@Generated(
+    value = "org.mapstruct.ap.MappingProcessor"
+)
+public class UserMapperImpl implements UserMapper {
+
+    private final ManagerMapper managerMapper = Mappers.getMapper(
+        ManagerMapper.class );
+    private final Validator validator = new Validator();
+
+    @Override
+    public BasicUserDTO convert(BasicUser user) throws ValidationException {
+        if ( user == null ) {
+            return null;
+        }
+
+        BasicUserDTOBuilder basicUserDTO = BasicUserDTO.builder();
+
+        basicUserDTO.id( validator.validateId( user.getId() ) );
+        basicUserDTO.name( user.getName() );
+
+        return basicUserDTO.build();
+    }
+
+    @Override
+    public BasicUser convert(BasicUserDTO userDTO) throws ValidationException {
+        if ( userDTO == null ) {
+            return null;
+        }
+
+        BasicUserBuilder basicUser = BasicUser.builder();
+
+        basicUser.id( validator.validateId( userDTO.getId() ) );
+        basicUser.name( userDTO.getName() );
+
+        return basicUser.build();
+    }
+    
+    ...
+}
+```
+
+MapStruct has automatically set the id of the mapper objects with the result of the `Validator` instance. It has added a `throws` clause for the method as well.
+
+## Mapper Retrieval Strategies
```

Comment:
I would move this section to the end because it's kind of a "mental break" to the other sections.

Reply:
I have moved this below. Please let me know if that looks suitable.